corehq/apps/userreports/tasks.py

```diff
@@ -303,119 +303,144 @@ def _queue_chunk(indicators):
+            assert indicator.doc_type == indicators[0].doc_type
 
-        for i in indicators:
-            assert i.domain == first_indicator.domain
-            assert i.doc_type == first_indicator.doc_type
+            for config_id in indicator.indicator_config_ids:
+                indicators_by_config_id[config_id].append(indicator)
+        return indicators_by_config_id
 
-        indicator_by_doc_id = {i.doc_id: i for i in indicators}
-        doc_store = get_document_store_for_doc_type(first_indicator.domain, first_indicator.doc_type)
-        indicator_config_ids = first_indicator.indicator_config_ids
-        related_docs_to_rebuild = set()
+    timer = TimingContext()
+    lock_keys = [
+        get_async_indicator_modify_lock_key(indicator_id)
+        for indicator_id in indicator_doc_ids
+    ]
+    with CriticalSection(lock_keys):
+        all_indicators = AsyncIndicator.objects.filter(
+            doc_id__in=indicator_doc_ids
+        )
+        if not all_indicators:
+            return
 
+        doc_store = get_document_store_for_doc_type(
+            all_indicators[0].domain, all_indicators[0].doc_type
+        )
+        related_doc_ids = set()
+        failed_indicators = set()
+        indicators_by_config_id = index_by_config_id(all_indicators)
         with timer:
-            for doc in doc_store.iter_documents(list(indicator_by_doc_id.keys())):
-                indicator = indicator_by_doc_id[doc['_id']]
-                successfully_processed, to_remove, rebuild_related_docs = _save_document_helper(indicator, doc)
-                if rebuild_related_docs:
-                    related_docs_to_rebuild = related_docs_to_rebuild.union(icds_get_related_docs_ids(doc['_id']))
-                if successfully_processed:
-                    processed_indicators.append(indicator.pk)
+            for config_id, indicators in six.iteritems(indicators_by_config_id):
+                doc_ids = [i.doc_id for i in indicators]
+                indicator_ids = [i.pk for i in indicators]
+                try:
+                    config = _get_config(config_id)
+                except (ResourceNotFound, StaticDataSourceConfigurationNotFoundError):
+                    celery_task_logger.info("{} no longer exists, skipping".format(config_id))
+                    # remove because the config no longer exists
+                    _mark_config_to_remove(config_id, indicator_ids)
+                    continue
+                except ESError:
+                    celery_task_logger.info("ES errored when trying to retrieve config")
+                    failed_indicators = failed_indicators.union(indicators)
+                    continue
+                adapter = None
+                docs = _memoized_get_docs(doc_ids, doc_store)
+                try:
+                    adapter = get_indicator_adapter(config, can_handle_laboratory=True)
+                    # Todo; evaluationcontext and reset_iteration?
```

Comment:
@emord I am not sure why evaluationcontext is [needed](https://github.com/dimagi/commcare-hq/pull/20743/files#diff-1f15a4ea1fec5c4ee5fa155d4d80c3d7L394). I have removed it here since it gets added in the adapter.get_all_values. There is also probably no need for `reset_iteration` since the evaluationcontext is added per save. Let me know if otherwise?